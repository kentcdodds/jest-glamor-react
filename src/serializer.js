const {replaceSelectors} = require('./replace-selectors')
const {
  getNodes,
  markNodes,
  getSelectors,
  getStylesAndAllSelectors,
} = require('./utils')

function createSerializer(styleSheet, classNameReplacer) {
  function test(val) {
    return (
      val &&
      !val.serializedWithJestGlamorReact &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (val instanceof HTMLElement && !isBeingSerialized(val)))
    )
  }

  function print(val, printer) {
    const nodes = getNodes(val)
    markNodes(nodes)
    const selectors = getSelectors(nodes)
    const {styles, allSelectors} = getStylesAndAllSelectors(
      selectors,
      styleSheet,
    )
    const printedVal = printer(val)
    if (styles) {
      return replaceSelectors(
        allSelectors,
        styles,
        printedVal,
        classNameReplacer,
      )
    } else {
      return printedVal
    }
  }

  return {test, print}
}

function isBeingSerialized(node) {
  let currentNode = node

  while (currentNode) {
    if (currentNode.serializedWithJestGlamorReact) {
      return true
    }
    currentNode = currentNode.parentNode
  }
  return false
}

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getGlamorStyleSheet = () => require('glamor').styleSheet
const glamorSerializer = createSerializer(getGlamorStyleSheet)
createSerializer.test = glamorSerializer.test
createSerializer.print = glamorSerializer.print

module.exports = createSerializer
