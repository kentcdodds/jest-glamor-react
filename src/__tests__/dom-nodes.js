import React from 'react'
import ReactDOM from 'react-dom'
import * as glamor from 'glamor'
import serializer from '../serializer'

expect.addSnapshotSerializer(serializer)

function Wrapper(props) {
  const className = glamor.css({
    padding: '4em',
    background: 'papayawhip',
  })
  return <section className={`${className}`} {...props} />
}

test('works when the root element does not have styles', () => {
  const div = render(
    <div>
      <Wrapper />
    </div>,
  )

  expect(div).toMatchSnapshot()
})

test(`doesn't mess up stuff that does't have styles`, () => {
  expect(render(<div />)).toMatchSnapshot()
})

test(`doesn't mess up stuff when styles have a child selector`, () => {
  const style = glamor.css({
    '> div': {
      display: 'inline-block',
    },
  })

  const div = render(
    <div>
      <span {...style} />
    </div>,
  )

  expect(div).toMatchSnapshot()
})

const generalTests = [
  {
    title: 'data attributes',
    styles: {
      backgroundColor: 'rebeccapurple',
      margin: 2,
    },
  },
  {
    title: 'media queries',
    styles: {
      fontSize: 24,
      margin: 12,
      '@media (max-width: 641px)': {
        fontSize: 20,
        margin: 10,
      },
    },
  },
  {
    title: 'pseudo elements',
    styles: {
      '& button': {
        color: 'green',
      },
    },
  },
  {
    title: 'pseudo states',
    styles: {
      ':hover': {
        backgroundColor: 'blue',
      },
    },
  },
  {
    title: 'Appended class',
    styles: {
      '&.button': {
        color: 'green',
      },
    },
  },
  {
    title: 'Child selector',
    styles: {
      '> div': {
        display: 'inline-block',
      },
    },
  },
]

generalTests.forEach(({title, styles}, index) => {
  test(title, () => {
    const div = render(<div {...glamor.css(styles)} />)
    expect(div).toMatchSnapshot(
      `${index + 1}. general tests: ${title} - ${JSON.stringify(styles)}`,
    )
  })
})

function render(ui) {
  const div = document.createElement('div')
  ReactDOM.render(ui, div)
  return div.children[0]
}
