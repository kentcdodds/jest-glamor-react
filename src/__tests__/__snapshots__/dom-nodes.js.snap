// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appended class: 5. general tests: Appended class - {"&.button":{"color":"green"}} 1`] = `
.css-0,
[data-css-1].button {
  color: green;
}

<div
  data-css-1=""
/>
`;

exports[`Child selector: 6. general tests: Child selector - {"> div":{"display":"inline-block"}} 1`] = `
.css-0> div,
[data-css-0]> div {
  display: inline-block;
}

<div
  data-css-0=""
/>
`;

exports[`data attributes: 1. general tests: data attributes - {"backgroundColor":"rebeccapurple","margin":2} 1`] = `
.css-0,
[data-css-0] {
  background-color: rebeccapurple;
  margin: 2px;
}

<div
  data-css-0=""
/>
`;

exports[`doesn't mess up stuff that does't have styles 1`] = `<div />`;

exports[`doesn't mess up stuff when styles have a child selector 1`] = `
.css-0> div,
[data-css-0]> div {
  display: inline-block;
}

<div>
  <span
    data-css-0=""
  />
</div>
`;

exports[`media queries: 2. general tests: media queries - {"fontSize":24,"margin":12,"@media (max-width: 641px)":{"fontSize":20,"margin":10}} 1`] = `
.css-0,
[data-css-0] {
  font-size: 24px;
  margin: 12px;
}

@media (max-width: 641px) {
  .css-0,
  [data-css-0] {
    font-size: 20px;
    margin: 10px;
  }
}

<div
  data-css-0=""
/>
`;

exports[`pseudo elements: 3. general tests: pseudo elements - {"& button":{"color":"green"}} 1`] = `
.css-0 button,
[data-css-0] button {
  color: green;
}

<div
  data-css-0=""
/>
`;

exports[`pseudo states: 4. general tests: pseudo states - {":hover":{"backgroundColor":"blue"}} 1`] = `
.css-0,
[data-css-1]:hover {
  background-color: blue;
}

<div
  data-css-1=""
/>
`;

exports[`works when the root element does not have styles 1`] = `
.css-0,
[data-css-0] {
  padding: 4em;
  background: papayawhip;
}

<div>
  <section
    class="css-0"
  />
</div>
`;
